### ISSUE TITLE
Support separate bridge networks and linking

### ISSUE THREAD

**Issue by armab** [drone/drone#1268]
#### Preamble

In Drone `v0.3` it was possible to access linked services by hostname.
In `v0.4` it was replaced with logic that [all services are available via `127.0.0.1`](http://readme.drone.io/build/services.html) only.

For most of the users `v0.4` approach is very useful and saves the time, also it's the same as we have in other well-known CI systems. However there are more complex configurations.
#### The problem

For exampe when not only `mongo`/`mysql`/another database is needed and there are many docker containers talk to each other by host and do complex things with ssh/scp, which is not possible on `127.0.0.1` due to port conflicts.
The strongest reason is to be better compatible with complex configurations defined in [`docker-compose.yml`](https://docs.docker.com/compose/), which uses the same approach with networking. Related issue https://github.com/drone/drone/issues/906 shows that it's pretty popular way.

You can have complex orchestration rules in `docker-compose` and port that logic into Drone CI. Defnitely cool!
#### Proposal

Add an option in `compose` section to access linked services by hostname.

In that case in `.drone.yml` config: 

``` yaml
compose:
  cache:
    image: redis
  database:
    image: postgres
```

`cache` and `database` would be hostnames injected into `/etc/hosts` of all containers.

---

I believe it would be super-helpful to find the way to be more compatible with `docker-compose` configurations, ideally to make CI run and `docker-compose` runs reproducible.

**Comment by bradrydzewski** [comment]
This will also be useful for individuals wanting to run Selenium tests against multiple browsers. The Chrome and Firefox selenium image both use port 4444. Since all builds share the same network this will cause a port conflict.

I propose that when explicitly setting `bridge` in the `.drone.yml` we avoid the default shared network and instead setup a bridge network:

```
compose:
  chrome:
    image: selenium/standalone-chrome
    net: bridge
```

We would then need to support container links. This is currently disabled because container names and links would need to be altered at runtime to ensure unique names, to avoid conflict with other running builds on the machine (especially if two builds run at once).

**Comment by dennybaa** [comment]
+1 to this feature. Can't use drone 0.4 without it (

**Comment by jakubsacha** [comment]
I have a question here. docker-compose is very good tool and i'm not sure is there a reason to reinvent the wheel.

With syntax like this:
```
compose:
   source: docker-compose.yml
   tests_container: primary_machine_name
   commands:
      - go get
      - go build
```

The concept is that having tests_machine id, we could read docker-compose.yml, overwrite tests_machine config with needed entrypoint to execute tests and spawn everything using docker-compose functionality. This would be very nice because:
- keeps consistency between docker-compose.yml and .drone.yml
- keeps drone open for new features coming with docker-compose

The only one tricky part would be to overwrite container part with predefined commands.

Is something like this possible at all?

**Comment by bradrydzewski** [comment]
@jakubsacha see issue https://github.com/drone/drone/issues/906 for importing a docker compose file as part of the yaml. This is definitely planned as a convenience to auto-populate the compose section. It would look something like this:

```
-compose:
-  database:
-    image: mysql
+compose: path/to/docker-compose.yml
```

a few things to consider ...

**First** is that many teams using Drone are not using Docker for deployment. We don't want to enforce an advanced level of Docker knowledge to use Drone.

**Second** is that many people building libraries are not using Docker. If you assume that 9 of 10 repositories on GitHub are libraries, and not applications, we don't want to enforce Docker knowledge to use Drone.

**Third** is that Drone has much more complex orchestration requirements than provided by `libcompose`, including running sequential, parallel and chained tasks, managing plugins and more. The `compose` section simply re-uses Drone's internal orchestration engine. We compile the Yaml to an AST representation, and we are good to go!

Using `libcompose` for the the `compose` section is certainly be possible, however, it wouldn't eliminate the need for Drone's internal orchestration engine. It would likely require a lot of code to have two engines working together, with security and sanization that we have today.

If you consider that supporting a field in Docker compose is **usually** around three lines of code [1] and that we would still need our own internal engine, I think using `libcompose` is more work than maintaining feature parity.

Linking and named containers are the exception, however, and will require a bit more code. I personally don't think it is enough code to justify the concerns with `libcompose`. I definitely appreciate the suggesting, and would encourage you to dig deeper into the orchestration code and let me know if you agree with my assessment.

[1] https://github.com/drone/drone-exec/commit/6b5e5a1118c8edab59a9aca8efc035bad80ca2cc

**Comment by sylus** [comment]
Are there any known workarounds / hacks to get a drone container build environment to see another container (gitlab) in a custom network?

**Comment by bradrydzewski** [comment]
@sylus I recommend asking this question in discuss.drone.io since there isn't enough information to debug and I don't want to sidetrack this particular issue, but I am happy to help. When you start a thread please provide all applicable information mentioned in https://discuss.drone.io/t/how-to-ask-for-help/188

**Comment by sylus** [comment]
Thanks for the assistance I have moved my question over to discuss.drone.io. ^_^

**Comment by bradrydzewski** [comment]
Since this is coming up on twitter and I can't type it in 140 characters, I'm going to detail how we can implement __v2__ Compose networking in the 0.5 or later versions of Drone.

As everyone known by now drone defaults to kubernetes pod-style networking, meaning all containers in the build pod share the same network. We have not picked one over the other. It is possible to provide both. It just happens that we had to start somewhere, and decided to pick the approach that we thought would be useful to the largest group of developers today.

### Pod Networking

Drone uses what is typically referred to as Pod networking. When Drone starts a build, all containers that are part of that build share the same networking using the Ambassador pattern. This allows all services to access eachother on localhost.

So even though the yaml is nearly identical to Docker compose (as seen below) the default network implementation is quite different.

```
compose:
  database:
    image: postgres
```

### Compose Networking

Compose networking (specifically v2, we will not implement v1) creates a default network that is shared among all containers. The default network is implied, but if we had to create it, it would look something like this:

```
pipeline:
  build:
    image: golang
    commands: [ go build, go test ]

compose:
  database:
    image: postgres

networks:
  default:
    driver: bridge
```

Compose networking will expose Postgres by hostname. The hostname is the container name, which in this case is `database`. This means the Postgres database will be available at `postgres://database:5432` to our build containers

So first of all, how do we instruct Drone to use Compose networking vs Pod networking? My proposal is that you simply include the `networks` section in the `.drone.yml`. The existence of the `networks` section will instruct Drone to create the custom bridge network and use it for all containers in the build.

Some good news -- the Yaml parser already is capable of parsing the network section.

I realize this deviates slightly from Docker Compose by requiring explicit network definition, however, until >= 50% of developers are using Docker Compose we are going to default to Pod-style networking. If and when Docker Compose becomes the norm we can change this. Until such a time this point is not up for debate (don't mean to sound strict here, but I don't have time or energy to debate this and my mind is made up).

### Network Conflicts

This is of course only part of the implementation. We need to make sure that user-defined networks have unique naming. Let's say a user includes a network in the Yaml like this:

```
networks:
  foo:
    driver: bridge
```

We cannot just do a `docker network create foo` because if two builds run on the same server at or around the same time and try to use the same name we will run into issues. Luckily Docker let's use give aliases to network names. We could call the network `foo-${RANDOM}` but map it into the container as `foo`. Problem solved!

We just need to write the code that goes through and changes any references to `foo` network to `foo-${RANDOM}`. This can be a transform function in our Yaml packages (transforms are used to transform the Yaml during the parsing stage, in case you were wondering).

### Summary

So in summary this is my proposal for v2 Compose networking in Drone. It is something I'm interested in implementing but is not something I will likely have time to implement until after this summer. If anyone is interested in using this as a blueprint for implementation, and would like to submit a pull request, please get in touch in our [gitter](https://gitter.im/drone/drone) room.

**Comment by bradrydzewski** [comment]
fixed in master. drone now uses docker-compose networking with host names. This will be available when the next version is officially released
